src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+        return .{ .value = addr };
+    }
+
+    /// Creates a physical address that points to `0`.
+    pub fn zero() PhysAddr {
+        return .{ .value = 0 };
+    }
+
+    /// Convenience method for checking if a physical address is null.
+    pub fn isNull(self: PhysAddr) bool {
+        return self.value == 0;
+    }
+
+    /// Aligns the physical address upwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignUp(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignForward(self.value, alignment) };
+    }
+
+    /// Aligns the physical address downwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignDown(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignBackward(@intCast(usize, self.value), alignment) };
+    }
+
+    /// Checks whether the physical address has the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn isAligned(self: PhysAddr, alignment: usize) bool {
+        return std.mem.isAligned(self.value, alignment);
+    }
+
+    test {
+        std.testing.refAllDecls(@This());
+        try std.testing.expectEqual(@bitSizeOf(u64), @bitSizeOf(PhysAddr));
+        try std.testing.expectEqual(@sizeOf(u64), @sizeOf(PhysAddr));
+    }
+};
+
+// ~~~ [ zig-x86_64/src/stuctures/paging/page.zig ] ~~~
+
+const SIZE_4KiB_STR: []const u8 = "4KiB";
+const SIZE_2MiB_STR: []const u8 = "2MiB";
+const SIZE_1GiB_STR: []const u8 = "1GiB";
+
+pub const PageSize = enum {
+    Size4KiB,
+    Size2MiB,
+    Size1GiB,
+
+    pub fn bytes(self: PageSize) u64 {
+        return switch (self) {
+            .Size4KiB => 4096,
+            .Size2MiB => 4096 * 512,
+            .Size1GiB => 4096 * 512 * 512,
+        };
+    }
+
+    pub fn sizeString(self: PageSize) []const u8 {
+        return switch (self) {
+            .Size4KiB => SIZE_4KiB_STR,
+            .Size2MiB => SIZE_2MiB_STR,
+            .Size1GiB => SIZE_1GiB_STR,
+        };
+    }
+
+    pub fn isGiantPage(self: PageSize) bool {
+        return self == .Size1GiB;
+    }
+};
+
+// ~~~ [ zig-x86_64/src/instructions/tlb.zig ] ~~~
+
+/// Structure of a PCID. A PCID has to be <= 4096 for x86_64.
+pub const Pcid = packed struct {
+    value: u12,
+
+    /// Create a new PCID
+    pub fn init(pcid: u12) Pcid {
+        return .{
+            .value = pcid,
+        };
+    }
+};
+
+// ~~~ [ zig-bitjuggle/bitjuggle.zig ] ~~~
+
+/// Obtains the `number_of_bits` bits starting at `start_bit`
+/// Where `start_bit` is the lowest significant bit to fetch
+///
+/// ```zig
+/// const a: u8 = 0b01101100;
+/// const b = getBits(a, 2, 4);
+/// try testing.expectEqual(@as(u4,0b1011), b);
+/// ```
+pub fn getBits(target: anytype, comptime start_bit: comptime_int, comptime number_of_bits: comptime_int) std.meta.Int(.unsigned, number_of_bits) {
+    const TargetType = @TypeOf(target);
```

Comment:
Runtime type resolving? Is this how they solve generics in zig?

Reply:
Yep, this is how generics is implemented.

ref: https://ziglang.org/documentation/master/#Compile-Time-Parameters